internal/app/machined/pkg/controllers/runtime/pcirebind.go

```diff
@@ -0,0 +1,224 @@
+					return fmt.Errorf("error checking bound driver for device with id: %s, %w", cfg.TypedSpec().VendorDeviceID, err)
+				}
+
+				if boundDriver != cfg.TypedSpec().TargetDriver { // TODO: this will fail if the module is not loaded
+					return fmt.Errorf("error binding device with id: %s to target driver: %s, bound driver: %s", cfg.TypedSpec().VendorDeviceID, cfg.TypedSpec().TargetDriver, boundDriver)
+				}
+
+				logger.Info("PCI device bound to target driver", zap.String("vendorDeviceID", cfg.TypedSpec().VendorDeviceID), zap.String("targetDriver", cfg.TypedSpec().TargetDriver))
+
+				if err := safe.WriterModify[*runtime.PCIRebindStatus](ctx, r, runtime.NewPCIRebindStatus(cfg.TypedSpec().Name), func(res *runtime.PCIRebindStatus) error {
+					res.TypedSpec().Name = cfg.TypedSpec().Name
+					res.TypedSpec().VendorDeviceID = cfg.TypedSpec().VendorDeviceID
+					res.TypedSpec().HostDriver = c.initialVendorDeviceToBoundDriver[cfg.TypedSpec().VendorDeviceID]
+					res.TypedSpec().TargetDriver = cfg.TypedSpec().TargetDriver
+
+					return nil
+				}); err != nil {
+					return fmt.Errorf("error updating PCI rebind status: %w", err)
+				}
+
+				c.vendorDeviceToBoundDriver[cfg.TypedSpec().VendorDeviceID] = cfg.TypedSpec().TargetDriver
+				touchedIDs[cfg.TypedSpec().VendorDeviceID] = struct{}{}
+			}
+		}
+
+		// cleanup any PCI devices that were not touched in the current run.
+		for vendorDeviceID, boundDriver := range c.vendorDeviceToBoundDriver {
+			if _, ok := touchedIDs[vendorDeviceID]; !ok {
+				if err := c.handlePCIBindToHost(vendorDeviceID, boundDriver); err != nil {
+					return err
+				}
+
+				logger.Info("PCI device bound to host driver", zap.String("vendorDeviceID", vendorDeviceID), zap.String("hostDriver", boundDriver))
+
+				delete(c.vendorDeviceToBoundDriver, vendorDeviceID)
+				delete(c.initialVendorDeviceToBoundDriver, vendorDeviceID)
+			}
+		}
+
+		if err := safe.CleanupOutputs[*runtime.PCIRebindStatus](ctx, r); err != nil {
+			return err
+		}
+	}
+}
+
+// handlePCIBindToTarget binds PCI device to a target driver and unbinds it from the host driver.
+func (c *PCIRebindController) handlePCIBindToTarget(vendorDeviceID, targetDriver string) error {
+	if err := handleDriverOverride(vendorDeviceID, targetDriver); err != nil {
+		return err
+	}
+
+	// Unbind device from the host driver.
+	// in some cases, the device may not be bound to any driver, so we ignore the error.
+	if err := os.WriteFile(fmt.Sprintf(driverUnbindPath, vendorDeviceID), []byte(vendorDeviceID), 0o200); err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("error unbinding device with id: %s, %w", vendorDeviceID, err)
+	}
+
+	return handleDriverProbe(vendorDeviceID)
+}
+
+// handlePCIBindToHost unbinds PCI device from a target driver and binds it to the host driver.
+func (c *PCIRebindController) handlePCIBindToHost(vendorDeviceID, hostDriver string) error {
+	if err := handleDriverOverride(vendorDeviceID, hostDriver); err != nil {
```

Comment:
I guess we typically want to send just "" there and it'll remove the override. Not 100% sure, but try, this could perhaps remove the need to keep a map of old drivers.

Reply:
https://www.kernel.org/doc/Documentation/ABI/testing/sysfs-bus-pci

> When specified, only a driver with a name matching the value written
		to driver_override will have an opportunity to bind to the
		device.  The override is specified by writing a string to the
		driver_override file (echo pci-stub > driver_override) and
		may be cleared with an empty string (echo > driver_override).
		This returns the device to standard matching rules binding.

...

> This also allows devices to
		opt-out of driver binding using a driver_override name such as
		"none".
		
So I guess clearing the file should restore binding by vendor ID